doc/design.md

```diff
@@ -0,0 +1,153 @@
+# Design of the Error Code Crawler Maven Plugin
+
+This file contains the design from the Error Code Crawler Maven Plugin (ECM). The design fulfills the requirements we described in [requirements.md](requirements.md).
+
+## Architecture
+
+`arch~components~1`
+
+ECM consists of the following main components:
+
+* A crawler: Components that reads extracts the error code declarations from Java code into a Java class structure
+* A validator: Component that validates the declarations
+* A config parser: Component that parses the `error_code_config.yml`
+* A report writer (**missing**): Component that writes the error code report
+
+Needs: dsn
+
+## The Crawler
+
+`dsn~error-declaration-crawler~1`
+
+For implementing the error declaration crawler we decided to use [Spoon](https://spoon.gforge.inria.fr/). This library can create an Abstract Syntax Tree (AST) from Java source code. The error declaration crawler then goes over the AST and searches for invocations of the [error-reporting](https://github.com/exasol/error-reporting-java/) API.
+
+In contrast to using regular expression matching, this approach has the following advantages:
+
+* The crawler gets the source code in a unified form. Whitespace and line breaks can't cause errors
+* The crawler can handle string concatenation can constants in error messages
+
+One could want to support string concatenation with variables or crawl error declarations like:
+
+```java
+class Test {
+    void test() {
+        var builder = ExaError.messageBuilder("E-TEST-1");
+        if (x) {
+            builder.message("Message 1");
+        } else {
+            builder.message("Message 2");
+        }
+        builder.toString();
+    }
+}
+```
+
+Supporting this is, however, not desirable since it make it impossible to determine the error message using static code analysis, since the error message will only be determined at runtime.
+
+Covers:
+
+* `arch~components~1`
+
+Needs: impl, utest
+
+## The Validator
+
+`dsn~validator~1`
+
+The validator consists of multiple sub-validators for the different aspects. The validator executes them.
+
+Covers:
+
+* `arch~components~1`
+
+Needs: impl, itest
+
+### Duplication Validator
+
+`dsn~duplication-validator~1`
+
+The duplication validator check that each error tag is unique.
+
+The requirements only require that the whole tag is unique. We decided however, to be more strict and also not allow tags that only differ by the number So this implementation considers `E-TEST-1` and `F-TEST-1` as duplicate. This approach has two advantages:
+
+* Users can simply change the type of the code without searching for a new number
```

Comment:
I don't understand what is meant here?

Reply:
I tried to clearify

Reply:
ah ok, with users you mean developers, here, maybe use developers instead users here

Reply:
done